sharding/utils/marshal.go

```diff
@@ -45,149 +48,140 @@ func ConvertFromRawBlob(blob *RawBlob, i interface{}) error {
-	// to the data chunks. The data chunks are separated into sets of 31.
-	for i := int64(1); i <= chunksNumber; i++ {
-		tempBody = append(tempBody,
-			append(indicatorByte,
-				cb.data[(i-1)*chunkDataSize:i*chunkDataSize]...)...)
-	}
-	// Appends indicator bytes to terminal-chunks , and if the index of the chunk
-	// delimiter is non-zero adds it to the chunk. Also pads empty bytes to
-	// the terminal chunk.chunkDataSize*chunksNumber refers to the total
-	// size of the blob. finalchunkIndex refers to the index of the last data byte.
-	indicatorByte[0] = byte(terminalLength)
-	if cb.flags.skipEvmExecution {
-		indicatorByte[0] |= (1 << 7)
-	}
-	tempBody = append(tempBody,
-		append(indicatorByte,
-			cb.data[chunkDataSize*chunksNumber:length]...)...)
+				indicatorByte := byte(terminalLength)
+				// include skipEvm flag if true
+				if rawBlob.flags.skipEvmExecution {
+					indicatorByte = indicatorByte | skipEvmBits
+				}
 
-	emptyBytes := make([]byte, (chunkDataSize - terminalLength))
-	tempBody = append(tempBody, emptyBytes...)
+				returnData = append(returnData, indicatorByte)
+			}
 
-	return tempBody, nil
+			// append data bytes
+			chunkStart := j * int(chunkDataSize)
+			chunkEnd := chunkStart + terminalLength
+			blobSlice := rawBlob.data[chunkStart:chunkEnd]
+			returnData = append(returnData, blobSlice...)
+
+			// append filler bytes, if necessary
+			if terminalLength != int(chunkDataSize) {
+				numFillerBytes := numChunks*int(chunkDataSize) - len(rawBlob.data)
+				fillerBytes := make([]byte, numFillerBytes)
+				returnData = append(returnData, fillerBytes...)
+			}
+		}
+	}
 
+	return returnData, nil
 }
 
-// Serialize takes a set of blobs and converts them to a single byte array.
-func Serialize(rawblobs []*RawBlob) ([]byte, error) {
-	length := int64(len(rawblobs))
-
-	serialisedData := []byte{}
-
-	//Loops through all the blobs and serializes them into chunks
-	for i := int64(0); i < length; i++ {
-		data := *rawblobs[i]
-		refinedData, err := SerializeBlob(data)
-		if err != nil {
-			return nil, fmt.Errorf("Index %v: %v", i, err)
-		}
-		serialisedData = append(serialisedData, refinedData...)
-	}
+func isSkipEvm(indicator byte) bool {
+	return indicator&skipEvmBits>>7 == 1
+}
+func getDatabyteLength(indicator byte) int {
+	return int(indicator & dataLengthBits)
+}
 
-	return serialisedData, nil
+type SerializedBlob struct {
+	numNonTerminalChunks int
+	terminalLength       int
 }
 
 // Deserialize results in the byte array being deserialised and
 // separated into its respective interfaces.
 func Deserialize(data []byte) ([]RawBlob, error) {
+	chunksNumber := len(data) / int(chunkSize)
+	serializedBlobs := []SerializedBlob{}
+	numPartitions := 0
+
+	// first iterate through every chunk and identify blobs and their length
+	for i := 0; i < chunksNumber; i++ {
+		indicatorIndex := i * int(chunkSize)
+		databyteLength := getDatabyteLength(data[indicatorIndex])
+
+		// if indicator is non-terminal, increase blobSize by 31
```

Comment:
What does this comment mean? I don't see any increased blobSize by 31?

Reply:
Ah yeah, this comment's a little out of date. I'll update.